ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+)
+```
+
+*NOTE: This however will come with a heavy warning — the activity must be strictly thread-safe as
+the provided instance will be called from multiple threads (and possibly at the same exact time).*
+
+
+## Execute a child workflow
+
+`parent_workflow.rb`
+
+```ruby
+class ChildWorkflow < Temporal::Workflow
+  def execute(name)
+    "Hello, #{name}!"
+  end
+end
+
+class ParentWorkflow < Temporal::Workflow
+  def execute
+    workflow.execute_child_workflow(ChildWorkflow, 'parent', id: 'id-2')
+  end
+end
+```
+
+
+## Async workflow
+
+Ruby doesn't have a standard (or at least very popular) framework for async execution of the code.
+Because of this we can assume that most Ruby engineers using the SDK won't be immediately familiar
+with writing async code. Because of this we have decided to make async execution opt-in, keeping
+everything synchronous by default.
+
+We have also chosen an eager execution pattern where any `async` block is started immediately and
+performed until blocked before proceeding with the execution. This is a bit easier to get a hold of
+and eliminates the risk of forgetting to actually start an async block. However this means that
+we'll need to find a solution for unhandled promise rejections, which we'll cover separately in a
+more detailed workflow proposal.
+
+We will introduce a helper method `async` to execute a block of code asynchronously:
+
+```ruby
+require 'sum_activity'
+
+class AsyncWorkflow < Temporal::Workflow
+  def execute
+    activity_promise = async { workflow.execute_activity(SumActivity, 1, 2) }
+    workflow_promise = async { workflow.execute_child_workflow(ChildWorkflow, 'parent') }
+
+    # Block execution until both promises are resolved
+    activity_result, workflow_result = async.all(activity_promise, workflow_promise).await
+
+    activity_result # => 3
+    workflow_result # => 'Hello, parent!'
+    
+    promise = async { workflow.execute_activity(SumActivity, 3, 4) }
+    
+    # individual promises can also be waited on
+    result = promise.await
+    
+    result # => 7
+  end
+end
+```
+
+A bit more involved example:
+
+```ruby
+class AnotherAsyncWorkflow < Temporal::Workflow
+  def execute
+    promise_1 = async { wait_and_sum(1, 2, delay: 15) }
+    promise_2 = async { wait_and_sum(3, 4, delay: 10) }
+
+    # The same exact method can be called in a blocking fashion
+    wait_and_sum(5, 6, delay: 5) # => 11
+
+    # Block execution until any promise is resolved
+    async.any(promise_1, promise_2).await
+
+    # Returns immediately because it will already be resolved (shorter delay)
+    promise_2.result # => 7
+
+    # This would block until the promise is resolved (longer delay)
+    promise_1.result # => 3
+  end
+
+  private
+
+  def wait_and_sum(a, b, delay:)
+    workflow.sleep(delay)
+    workflow.execute_activity(SumActivity, a, b)
+  end
+end
+```
+
+Methods `async.all` and `async.any` return promises themselves, so they can be used just like the
```

Comment:
Note that Python combines these to just [wait](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait) and has an option for whether you want first completed or all. This is important because there's actually multiple wait styles:

* first success - first one finishes successfully or _all_ finish unsuccessfully (JS equiv to `Promise.any`)
* first completed - first one finishes or fails (JS equiv to `Promise.race`)
* all - all complete

You can keep the "any" and "all" separate, but you need to define how you handle a promise completed with a failure in the "any" case and you may need another "any" for the alternative approach

Reply:
`all` equivalent in JS is `Promise.all`, it also has `Promise.allSettled` that doesn't throw when a promise is rejected.
I find the JS approach simpler but don't have a strong preference, up to @antstorm to decide.

Reply:
+1 to different top level names. Easier to scan and read at a glance.

Reply:
Right now it's important to agree that `async.all` means "all finished" (completed or failed) and `async.any` means "first finished" (completed or failed).

I agree that top-level names are the best for readability and can be added later for more granular control. But overall this doesn't feel very important since we have `async.await_for` that can be used instead.